pkg/services/app/add.go

```diff
@@ -0,0 +1,503 @@
+		}
+		return nil
+	}
+
+	for _, manifest := range manifests {
+		if out, err := a.kube.Apply(manifest, params.Namespace); err != nil {
+			return errors.Wrap(err, fmt.Sprintf("could not apply manifest: %s", string(out)))
+		}
+	}
+
+	return nil
+}
+
+func (a *App) cloneRepo(url string, branch string, dryRun bool) error {
+	if dryRun {
+		return nil
+	}
+
+	url = sanitizeRepoUrl(url)
+
+	repoDir, err := ioutil.TempDir("", "user-repo-")
+	if err != nil {
+		return errors.Wrap(err, "failed creating temp. directory to clone repo")
+	}
+
+	_, err = a.git.Clone(context.Background(), repoDir, url, branch)
+	if err != nil {
+		return errors.Wrapf(err, "failed cloning user repo: %s", url)
+	}
+
+	return nil
+}
+
+func (a *App) writeAppYaml(basePath string, name string, manifest []byte) error {
+	manifestPath := filepath.Join(basePath, "apps", name, "app.yaml")
+
+	return a.git.Write(manifestPath, manifest)
+}
+
+func (a *App) writeAppGoats(basePath string, name string, clusterName string, manifests ...[]byte) error {
+	goatPath := filepath.Join(basePath, "targets", clusterName, name, fmt.Sprintf("%s-gitops-runtime.yaml", name))
+
+	goat := bytes.Join(manifests, []byte(""))
+	return a.git.Write(goatPath, goat)
+}
+
+func generateAppYaml(params AddParams) ([]byte, error) {
+	const appYamlTemplate = `---
+apiVersion: wego.weave.works/v1alpha1
+kind: Application
+metadata:
+  name: {{ .AppName }}
+spec:
+  path: {{ .AppPath }}
+  url: {{ .AppURL }}
+`
+	// Create app.yaml
+	t, err := template.New("appYaml").Parse(appYamlTemplate)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not parse app yaml template")
+	}
+
+	var populated bytes.Buffer
+	err = t.Execute(&populated, struct {
+		AppName string
+		AppPath string
+		AppURL  string
+	}{params.Name, params.Path, params.Url})
+	if err != nil {
+		return nil, errors.Wrap(err, "could not execute populated template")
+	}
+	return populated.Bytes(), nil
+}
+
+func generateResourceName(url string) string {
+	return strings.ReplaceAll(urlToRepoName(url), "_", "-")
+}
+
+func getOwnerFromUrl(url string) (string, error) {
+	parts := strings.Split(url, "/")
+	if len(parts) < 2 {
+		return "", fmt.Errorf("could not get owner from url %s", url)
+	}
+	return parts[len(parts)-2], nil
+}
+
+func urlToRepoName(url string) string {
+	return strings.TrimSuffix(filepath.Base(url), ".git")
+}
+
+func sanitizeRepoUrl(url string) string {
+	trimmed := ""
+
+	if !strings.HasSuffix(url, ".git") {
+		url = url + ".git"
+	}
+
+	sshPrefix := "git@github.com:"
+	if strings.HasPrefix(url, sshPrefix) {
+		trimmed = strings.TrimPrefix(url, sshPrefix)
+	}
+
+	httpsPrefix := "https://github.com/"
+	if strings.HasPrefix(url, httpsPrefix) {
+		trimmed = strings.TrimPrefix(url, httpsPrefix)
+	}
+
+	if trimmed != "" {
+		return "ssh://git@github.com/" + trimmed
+	}
+
+	return url
+}
+
+// NOTE: ready to save the targets automation in phase 2
+// func (a *App) writeTargetGoats(basePath string, name string, manifests ...[]byte) error {
```

Comment:
I know its not used yet, but can we avoid adding obfuscation with the term "Goats"?

Reply:
I agree, not a super fan, but struggled to find a good name that means something is `kustomization` or a `helmrelease`

Reply:
The Flux maintainers might have a word to summarize this class of "reconcilable" objects